File: packages/engine/src/core/index-store.ts
import { existsSync, mkdirSync, readFileSync, statSync } from 'node:fs'
import { join } from 'node:path'
import { writeFileAtomic } from './atomic'

/**
 * The derived index: local, disposable, never truth (record-index D1).
 *
 * Every cross-record question sofar can ask — which initiatives hold open
 * sessions, who else has this file, what cites this decision — currently costs
 * a sweep of the whole record. Measured, that is 18.9ms at 300 initiatives and
 * 64.1ms at 1000, against a shim budget of 100ms it already spends 63-67ms of.
 * Sweeping is the wrong shape: it re-derives from scratch what has not changed.
 *
 * The log makes a better shape available, and it is the property no
 * general-purpose index gets to assume: events are APPEND-ONLY and ulid
 * ordered, so a derivation over a prefix is permanently valid. Nothing is ever
 * invalidated, only extended. That turns maintenance into O(new events) and
 * makes a stale index a partial index rather than a wrong one.
 *
 * Three rules keep it honest, and they are the whole safety argument:
 *
 * 1. DERIVED, NEVER TRUTH. Truth is events.jsonl. Anything here can be deleted
 *    at any moment and rebuilt with no loss. No reader may return a wrong
 *    answer because the index was missing — only a slower one.
 * 2. LOCAL, NEVER COMMITTED. A committed index would conflict on every
 *    parallel append (it is not append-only, so the `merge=union` bargain in
 *    team-readiness T2 does not apply to it) and would arrive stale in every
 *    clone. The directory ignores itself, so no repo .gitignore has to know.
 * 3. VERSION-STAMPED. A schema change makes old files unreadable rather than
 *    subtly misread: any mismatch is a cold start, which costs one rebuild and
 *    never a wrong answer.
 */

/** Bump on ANY change to the on-disk shape. Old versions cold-start. */
export const INDEX_SCHEMA_VERSION = 3

/** How far one initiative's log has been consumed. */
export interface InitiativeCursor {
  /** Envelope id of the last consumed event — the authority. */
  id: string
  /**
   * Byte offset where that event's line STARTS, so a resume can seek instead
   * of rescanning. Storing the start rather than the end is what makes the
   * seek self-corroborating: the first line read back must carry `id`, and
   * when it does not, the offset is lying and the reader falls back to a full
   * read. Offsets lie more often than they look — an import, a rewrite, or a
   * restore leaves a perfectly valid number pointing at the wrong line.
   */
  offset: number
  /** Log size when the cursor was written; a shrink proves a rewrite. */
  size: number
  /**
   * Log mtime when the cursor was written. Paired with `size` it identifies
   * "this file has not been touched since I read it" from ONE stat, which is
   * what makes a quiet initiative cost a syscall instead of a read (v2).
   *
   * Neither half stands alone. An append always grows the file, so size alone
   * catches every append; a rewrite can preserve size, so mtime catches the
   * import/restore/checkout that size cannot see. Requiring both to match
   * means the reader only ever skips a file that is byte-identical AND
   * untouched — anything else falls through to the corroborated read below.
   * mtime is never TRUSTED here, only used as a difference detector, which is
   * the one thing it is honest about (warmth.ts rejects it as a warmth signal
   * for exactly the opposite reason: git checkout makes cold logs look fresh,
   * and looking fresh here only ever costs an extra read).
   */
  mtimeMs: number
  /**
   * Greatest event id consumed, which is NOT always the id at `offset`: the
   * fold replays in ulid order rather than file order (the convergent-fold
   * rule), and a union-merged log interleaves two branches' lines. Resuming
   * incrementally is only sound while every arriving id is above everything
   * already applied, so the pass compares against this and rebuilds when it
   * is not. Absent on a cursor written before this existed.
   */
  maxId?: string
  /**
   * Event ids voided by a `correction` in this log. The log is append-only,
   * but a correction is retroactive: it can void an event a previous pass
   * already applied, and an index that ignores it reports work the fold has
   * dropped. Kept per log so a void survives across refreshes — this record
   * alone carries 24 corrections, over session_started, session_ended,
   * decision_logged and command_run.
   */
  voided?: string[]
}

/** Size + mtime in one syscall — the cheapest question you can ask a log. */
export function logStat(logPath: string): { size: number; mtimeMs: number } | null {
  try {
    const s = statSync(logPath)
    return { size: s.size, mtimeMs: s.mtimeMs }
  } catch {
    return null
  }
}

export interface IndexMeta {
  version: number
  cursors: Record<string, InitiativeCursor>
}

export function indexDir(sofarDir: string): string {
  return join(sofarDir, '.index')
}

/**
 * Cursor file for the DEFAULT tier. Every tier names its own (record-index
 * 3.1): cursors say how far a tier has consumed each log, so sharing one file
 * across tiers would let whichever tier refreshed first advance the cursor
 * past events the others never saw. Separate files also keep the tiers
 * independently refreshable, which is the whole point of tiering — the hot
 * consumer must never pay to maintain a tier it does not read.
 */
export const DEFAULT_META_FILE = 'meta.json'

function metaPath(sofarDir: string, file: string): string {
  return join(indexDir(sofarDir), file)
}

/**
 * Create the index directory, self-ignoring.
 *
 * A `.gitignore` of `*` inside the directory hides the directory AND itself,
 * so the index never appears in `git status` and the repo's own .gitignore
 * never has to learn about it — which matters because that file belongs to the
 * user, and a tool editing it is a tool that will eventually clobber it.
 */
export function ensureIndexDir(sofarDir: string): string {
  const dir = indexDir(sofarDir)
  mkdirSync(dir, { recursive: true })
  const ignore = join(dir, '.gitignore')
  if (!existsSync(ignore)) writeFileAtomic(ignore, '*\n')
  return dir
}

/**
 * Read the index metadata, or null for "start cold".
 *
 * Null covers every failure identically — absent, unreadable, malformed,
 * wrong version, wrong shape — because every one of them means the same thing
 * to a caller: derive from the logs instead. Distinguishing them would only
 * tempt a reader into trusting a partially-valid file.
 */
export function readIndexMeta(sofarDir: string, file: string = DEFAULT_META_FILE): IndexMeta | null {
  try {
    const raw: unknown = JSON.parse(readFileSync(metaPath(sofarDir, file), 'utf8'))
    if (typeof raw !== 'object' || raw === null) return null
    const rec = raw as Record<string, unknown>
    if (rec.version !== INDEX_SCHEMA_VERSION) return null
    const cursors = rec.cursors
    if (typeof cursors !== 'object' || cursors === null) return null

    const clean: Record<string, InitiativeCursor> = {}
    for (const [slug, value] of Object.entries(cursors as Record<string, unknown>)) {
      const c = value as Record<string, unknown>
      if (typeof c?.id !== 'string' || c.id.length === 0) continue
      if (typeof c.offset !== 'number' || !Number.isInteger(c.offset) || c.offset < 0) continue
      if (typeof c.size !== 'number' || !Number.isInteger(c.size) || c.size < 0) continue
      if (typeof c.mtimeMs !== 'number' || !Number.isFinite(c.mtimeMs) || c.mtimeMs < 0) continue
      const cursor: InitiativeCursor = { id: c.id, offset: c.offset, size: c.size, mtimeMs: c.mtimeMs }
      if (typeof c.maxId === 'string' && c.maxId.length > 0) cursor.maxId = c.maxId
      if (Array.isArray(c.voided) && c.voided.every((v) => typeof v === 'string')) {
        cursor.voided = c.voided as string[]
      }
      clean[slug] = cursor
    }
    return { version: INDEX_SCHEMA_VERSION, cursors: clean }
  } catch {
    return null
  }
}

/** Persist metadata atomically. Silent on failure — the index is disposable. */
export function writeIndexMeta(sofarDir: string, meta: IndexMeta, file: string = DEFAULT_META_FILE): void {
  try {
    ensureIndexDir(sofarDir)
    writeFileAtomic(metaPath(sofarDir, file), `${JSON.stringify({ ...meta, version: INDEX_SCHEMA_VERSION })}\n`)
  } catch {
    // A record that cannot cache is a record that is merely slower. Never
    // let index maintenance fail an append, a hook, or a command.
  }
}

/**
 * Is this cursor still usable against the log as it stands now?
 *
 * Only the cheap, decisive checks live here — the log shrank, or vanished.
 * Both prove the cursor describes a file that no longer exists in that form
 * (an import, a rewrite, a restore from backup). Whether the offset points at
 * the right LINE is a question only the reader can answer, by checking the id
 * there, and it does exactly that.
 */
export function cursorPlausible(logPath: string, cursor: InitiativeCursor): boolean {
  const stat = logStat(logPath)
  return stat !== null && cursorUsable(stat, cursor)
}

/** The same test against a stat the caller already holds. */
export function cursorUsable(stat: { size: number }, cursor: InitiativeCursor): boolean {
  return stat.size >= cursor.size && cursor.offset <= stat.size
}

/**
 * Is this log byte-for-byte the file the cursor was written against?
 *
 * The fast path's entire justification, and the reason it is safe: an append
 * changes `size`, and a rewrite that preserves size changes `mtimeMs`. When
 * both match there is nothing to read, so the reader skips the file entirely
 * — measured at 1000 initiatives, that is the difference between 13ms of tail
 * reads and 1.8ms of stats on a path with a 100ms end-to-end budget.
 */
export function logUntouched(stat: { size: number; mtimeMs: number }, cursor: InitiativeCursor): boolean {
  return stat.size === cursor.size && stat.mtimeMs === cursor.mtimeMs
}

/** Read the derived payload stored under `name`, or null to start cold. */
export function readIndexFile<T>(sofarDir: string, name: string, guard: (v: unknown) => v is T): T | null {
  try {
    const raw: unknown = JSON.parse(readFileSync(join(indexDir(sofarDir), name), 'utf8'))
    return guard(raw) ? raw : null
  } catch {
    return null
  }
}

/** Write a derived payload. Silent on failure, like every write here. */
export function writeIndexFile(sofarDir: string, name: string, value: unknown): void {
  try {
    ensureIndexDir(sofarDir)
    writeFileAtomic(join(indexDir(sofarDir), name), `${JSON.stringify(value)}\n`)
  } catch {
    // See writeIndexMeta.
  }
}

File: crates/sofar-core/src/index_store.rs
//! The derived index (`core/index-store.ts`, record-index D1): local,
//! disposable, version-stamped, never truth. Cursors say how far a tier has
//! consumed each log; payload files hold what it derived. Every read failure
//! is a cold start and every write failure is silent.

use std::fs;
use std::path::Path;

use crate::atomic::write_file_atomic;
use crate::json::{self, Json, Object};
use crate::layout::Layout;

/// Bump on ANY change to the on-disk shape. Old versions cold-start.
pub const INDEX_SCHEMA_VERSION: f64 = 5.0;

/// Cursor file for the default tier.
pub const DEFAULT_META_FILE: &str = "meta.json";

/// How far one initiative's log has been consumed (`InitiativeCursor`).
#[derive(Debug, Clone, PartialEq)]
pub struct Cursor {
    /// Envelope id of the last consumed event.
    pub id: String,
    /// Byte offset where that event's line STARTS.
    pub offset: u64,
    /// Log size when the cursor was written.
    pub size: u64,
    /// Log mtime when the cursor was written (Node's `mtimeMs`).
    pub mtime_ms: f64,
    /// Greatest event id consumed.
    pub max_id: Option<String>,
    /// Event ids voided by a `correction` in this log, sorted.
    pub voided: Option<Vec<String>>,
}

impl Cursor {
    fn to_json(&self) -> Json {
        let mut o = Object::with_capacity(6);
        o.insert("id", Json::Str(self.id.clone()));
        #[allow(clippy::cast_precision_loss, reason = "file offsets fit f64")]
        {
            o.insert("offset", Json::Num(self.offset as f64));
            o.insert("size", Json::Num(self.size as f64));
        }
        o.insert("mtimeMs", Json::Num(self.mtime_ms));
        if let Some(max) = &self.max_id {
            o.insert("maxId", Json::Str(max.clone()));
        }
        if let Some(voided) = &self.voided {
            o.insert(
                "voided",
                Json::Arr(voided.iter().map(|v| Json::Str(v.clone())).collect()),
            );
        }
        Json::Obj(o)
    }

    fn from_json(value: &Json) -> Option<Cursor> {
        let c = value.as_obj()?;
        let id = c.get("id")?.as_nonempty_str()?.to_owned();
        let non_negative_int = |key: &str| -> Option<u64> {
            let n = c.get(key)?.as_f64()?;
            #[allow(
                clippy::cast_possible_truncation,
                clippy::cast_sign_loss,
                reason = "checked"
            )]
            (n.is_finite() && n >= 0.0 && n.fract() == 0.0).then_some(n as u64)
        };
        let offset = non_negative_int("offset")?;
        let size = non_negative_int("size")?;
        let mtime_ms = c.get("mtimeMs")?.as_f64()?;
        if !mtime_ms.is_finite() || mtime_ms < 0.0 {
            return None;
        }
        let max_id = c
            .get("maxId")
            .and_then(Json::as_nonempty_str)
            .map(str::to_owned);
        let voided = c.get("voided").and_then(Json::as_arr).and_then(|items| {
            items
                .iter()
                .map(|v| v.as_str().map(str::to_owned))
                .collect::<Option<Vec<String>>>()
        });
        Some(Cursor {
            id,
            offset,
            size,
            mtime_ms,
            max_id,
            voided,
        })
    }
}

/// Size + mtime in one syscall (`logStat`).
#[derive(Debug, Clone, Copy, PartialEq)]
pub struct LogStat {
    pub size: u64,
    pub mtime_ms: f64,
}

#[must_use]
pub fn log_stat(log_path: &Path) -> Option<LogStat> {
    let m = fs::metadata(log_path).ok()?;
    Some(LogStat {
        size: m.len(),
        mtime_ms: mtime_ms_of(&m),
    })
}

/// Node's `stat.mtimeMs`: `sec * 1e3 + nsec / 1e6` as a double.
#[must_use]
pub fn mtime_ms_of(m: &fs::Metadata) -> f64 {
    let Ok(t) = m.modified() else { return 0.0 };
    match t.duration_since(std::time::UNIX_EPOCH) {
        #[allow(clippy::cast_precision_loss, reason = "seconds fit f64")]
        Ok(d) => d.as_secs() as f64 * 1000.0 + f64::from(d.subsec_nanos()) / 1e6,
        Err(_) => 0.0,
    }
}

/// `cursorUsable`: the log has not shrunk below the cursor.
#[must_use]
pub fn cursor_usable(stat: LogStat, cursor: &Cursor) -> bool {
    stat.size >= cursor.size && cursor.offset <= stat.size
}

/// `logUntouched`: byte-for-byte the file the cursor was written against.
#[must_use]
#[allow(
    clippy::float_cmp,
    reason = "exact equality of a stored stat IS the contract"
)]
pub fn log_untouched(stat: LogStat, cursor: &Cursor) -> bool {
    stat.size == cursor.size && stat.mtime_ms == cursor.mtime_ms
}

/// `IndexMeta`: slug → cursor, in file order.
#[derive(Debug, Clone, Default, PartialEq)]
pub struct IndexMeta {
    pub cursors: Vec<(String, Cursor)>,
}

impl IndexMeta {
    #[must_use]
    pub fn get(&self, slug: &str) -> Option<&Cursor> {
        self.cursors.iter().find(|(s, _)| s == slug).map(|(_, c)| c)
    }

    pub fn set(&mut self, slug: &str, cursor: Cursor) {
        match self.cursors.iter_mut().find(|(s, _)| s == slug) {
            Some(slot) => slot.1 = cursor,
            None => self.cursors.push((slug.to_owned(), cursor)),
        }
    }

    pub fn remove(&mut self, slug: &str) -> bool {
        let before = self.cursors.len();
        self.cursors.retain(|(s, _)| s != slug);
        before != self.cursors.len()
    }
}

/// `readIndexMeta`: the cursors, or None for "start cold".
#[must_use]
pub fn read_index_meta(layout: &Layout, file: &str) -> Option<IndexMeta> {
    let text = fs::read(layout.index_dir().join(file)).ok()?;
    let Json::Obj(rec) = json::parse(&String::from_utf8_lossy(&text)).ok()? else {
        return None;
    };
    if rec.get("version") != Some(&Json::Num(INDEX_SCHEMA_VERSION)) {
        return None;
    }
    let cursors = rec.get("cursors")?.as_obj()?;
    let mut meta = IndexMeta::default();
    for (slug, value) in cursors.js_ordered() {
        if let Some(cursor) = Cursor::from_json(value) {
            meta.cursors.push((slug.to_owned(), cursor));
        }
    }
    Some(meta)
}

/// `writeIndexMeta`: atomic, silent on failure.
pub fn write_index_meta(layout: &Layout, meta: &IndexMeta, file: &str) {
    let mut cursors = Object::with_capacity(meta.cursors.len());
    for (slug, cursor) in &meta.cursors {
        cursors.insert(slug.clone(), cursor.to_json());
    }
    let mut o = Object::with_capacity(2);
    o.insert("version", Json::Num(INDEX_SCHEMA_VERSION));
    o.insert("cursors", Json::Obj(cursors));
    write_index_file(layout, file, &Json::Obj(o));
}

/// `readIndexFile`: the parsed payload under `name`, or None to start cold.
#[must_use]
pub fn read_index_file(layout: &Layout, name: &str) -> Option<Json> {
    let text = fs::read(layout.index_dir().join(name)).ok()?;
    json::parse(&String::from_utf8_lossy(&text)).ok()
}

/// `writeIndexFile`: `JSON.stringify(value) + '\n'`, atomic, silent on failure.
pub fn write_index_file(layout: &Layout, name: &str, value: &Json) {
    let Ok(dir) = layout.ensure_index_dir() else {
        return;
    };
    let mut text = json::stringify(value);
    text.push('\n');
    let _ = write_file_atomic(&dir.join(name), text.as_bytes());
}

/// `version === INDEX_SCHEMA_VERSION && initiatives is an object` (`isTierDisk`).
#[must_use]
pub fn tier_initiatives(disk: &Json) -> Option<&Object> {
    let o = disk.as_obj()?;
    if o.get("version") != Some(&Json::Num(INDEX_SCHEMA_VERSION)) {
        return None;
    }
    o.get("initiatives")?.as_obj()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn meta_round_trips_and_rejects_other_versions() {
        let dir = crate::testing::scratch_dir("index-store");
        let layout = Layout::new(&dir);
        let mut meta = IndexMeta::default();
        meta.set(
            "a",
            Cursor {
                id: "01ARZ3NDEKTSV4RRFFQ69G5FAV".into(),
                offset: 10,
                size: 20,
                mtime_ms: 1_700_000_000_123.456,
                max_id: None,
                voided: Some(vec!["x".into()]),
            },
        );
        write_index_meta(&layout, &meta, "meta-test.json");
        let text = fs::read_to_string(layout.index_dir().join("meta-test.json")).unwrap();
        assert_eq!(
            text,
            "{\"version\":5,\"cursors\":{\"a\":{\"id\":\"01ARZ3NDEKTSV4RRFFQ69G5FAV\",\"offset\":10,\"size\":20,\"mtimeMs\":1700000000123.456,\"voided\":[\"x\"]}}}\n"
        );
        assert_eq!(read_index_meta(&layout, "meta-test.json"), Some(meta));
        fs::write(
            layout.index_dir().join("meta-test.json"),
            "{\"version\":4,\"cursors\":{}}",
        )
        .unwrap();
        assert_eq!(read_index_meta(&layout, "meta-test.json"), None);
        fs::remove_dir_all(&dir).unwrap();
    }
}
